xlive/XLive/XUser/XUserContext.cpp

```diff
@@ -1,223 +1,213 @@
+	_context_presence_invite_only_game = 8,
+	_context_presence_network_in_game = 9
 };
 
-void update_player_count()
-{
-	s_network_session* session = nullptr;
-	if (NetworkSession::GetActiveNetworkSession(&session))
-	{
-		DiscordInterface::SetPlayerCountInfo(
-			session->membership[0].player_count, 
-			session->parameters[0].max_party_players);
-	}
-	else
-	{
-		DiscordInterface::SetPlayerCountInfo(0, 0);
-	}
-}
+extern void Check_Overlapped(PXOVERLAPPED pOverlapped);
+extern XSESSION_LOCAL_DETAILS sessionDetails;
 
-std::string getEnglishMapName()
+// 5289: XUserGetContext
+DWORD WINAPI XUserGetContext(DWORD dwUserIndex, XUSER_CONTEXT* pContext, PXOVERLAPPED  pOverlapped)
 {
-	wchar_t* p_englishMapName = Memory::GetAddress<wchar_t*>(0x97737C);
-	std::wstring englishMapName(p_englishMapName);
-	return  std::string(englishMapName.begin(), englishMapName.end());
+	return S_OK;
 }
 
-std::string getVariantName()
+void context_update_map_info_multiplayer(void)
 {
-	std::wstring variant = NetworkSession::GetGameVariantName();
-	variant = variant.substr(0, variant.find_last_not_of(L"\xE008\t\n ") + 1);
-	return std::string(variant.begin(), variant.end());
+	// Get map name and convert to utf8
+	wchar_t* wide_map_name = Memory::GetAddress<wchar_t*>(0x97737C);
+	utf8 map_name[512];
+	wchar_string_to_utf8_string(wide_map_name, map_name, sizeof(map_name));
+
+	// Get scenario name and convert to utf8
+	c_static_wchar_string260* scenario_path = &game_options_get()->scenario_path;
+	int32 index = scenario_path->last_index_of(L"\\");
+	const wchar_t* scenario_name_wide = &scenario_path->get_string()[index + 1];
+	utf8 scenario_name[MAX_PATH];
+	wchar_string_to_utf8_string(scenario_name_wide, scenario_name, sizeof(scenario_name));
+
+	// Update interface
+	discord_interface_set_map_name(scenario_name, map_name);
 }
 
-std::string gamemode_id_to_string(int id)
+void context_update_map_info_campaign(uint32 map_id, const utf8* scenario_name)
 {
-	std::string gamemode = "gamemode_unknown";
-	auto it = game_mode_list.find(id);
-	if (it != game_mode_list.end())
-		gamemode = it->second;
-	return gamemode;
-}
+	// Get map name
+	const char* map_name = NULL;
+	for (uint32 i = 0; i < NUMBEROF(g_context_campaign_map_info); ++i)
+	{
+		if (g_context_campaign_map_info[i].map_id == map_id)
+		{
+			map_name = g_context_campaign_map_info[i].map_name;
+			break;
+		}
+	}
 
-enum class ContextPresence {
-	mainmenu,
-	unknown,
-	server_browser,
-	singleplayer,
-	lobby,
-	results,
-	live_in_game,
-	public_game,
-	invite_only_game,
-	network_in_game
-};
+	// Convert map_name to utf8
+	if (map_name == NULL)
+	{
+		map_name = scenario_name;
+	}
 
-DWORD diff_level;
+	// Update interface
+	discord_interface_set_map_name(scenario_name, map_name);
+}
 
 // #5277: XUserSetContext
-DWORD WINAPI XUserSetContext(DWORD dwUserIndex, DWORD dwContextId, DWORD dwContextValue)
+DWORD WINAPI XUserSetContext(DWORD dwUserIndex, e_context_id dwContextId, DWORD dwContextValue)
```

Comment:
why `e_context_id` ? convert it inside the function definition, not in the prototype, this is an xlive API

Reply:
changed back